source/slang/slang-check-decl.cpp

```diff
@@ -3724,64 +3732,155 @@ namespace Slang
+        {
+            getSink()->diagnose(decl, Diagnostics::nonSetAccessorMustNotHaveParams);
+        }
+
+        // By default, the return type of an accessor is treated as
+        // the type of the abstract storage location being accessed.
+        //
+        // A `ref`  accessor currently relies on this logic even though
+        // it isn't quite correct, because we don't have support
+        // for by-reference return values today. This is a non-issue
+        // for now because we don't support user-defined `ref`
+        // accessors yet.
+        //
+        // TODO: Once we can support the by-reference return value
+        // correctly *or* we can move to something like a coroutine-based
+        // `modify` accessor (a la Swift), we should split out
+        // handling of `RefAccessorDecl` and only use this routine
+        // for `GetterDecl`s.
+        //
+        decl->returnType.type = _getAccessorStorageType(decl);
     }
 
     void SemanticsDeclHeaderVisitor::visitSetterDecl(SetterDecl* decl)
     {
         // Make sure to invoke the common checking logic for all accessors.
-        visitAccessorDecl(decl);
+        _visitAccessorDeclCommon(decl);
+
+        // A `set` accessor always returns `void`.
+        //
+        decl->returnType.type = getASTBuilder()->getVoidType();
 
+        // A setter always receives a single value representing
+        // the new value to set into the storage.
+        //
+        // The user may declare that parameter explicitly and
+        // thereby control its name, or they can declare no
+        // parmaeters and allow the compiler to synthesize one
+        // names `newValue`.
+        //
+        ParamDecl* newValueParam = nullptr;
         auto params = decl->getParameters();
-        ParamDecl* param = nullptr;
         if( params.getCount() >= 1 )
         {
-            param = params.getFirst();
+            // If the user declared an explicit parameter
+            // then that is the one that will represent
+            // the new value.
+            //
+            newValueParam = params.getFirst();
+
             if( params.getCount() > 1 )
             {
-                // TODO: diagnose it!
+                // If the user declared more than one explicit
+                // parameter, then that is an error.
+                //
+                getSink()->diagnose(params[1], Diagnostics::setAccessorMayNotHaveMoreThanOneParam);
             }
         }
         else
         {
-            param = m_astBuilder->create<ParamDecl>();
-            param->nameAndLoc.name = getName("newValue");
-            param->nameAndLoc.loc = decl->loc;
+            // If the user didn't declare any explicit parameters,
+            // then we create an implicit one and add it into
+            // the AST.
+            //
+            newValueParam = m_astBuilder->create<ParamDecl>();
+            newValueParam->nameAndLoc.name = getName("newValue");
```

Comment:
Should we stay consistent with C# and just use `value` instead?

Reply:
I'd be fine with changing this default down the line, but I won't do it in this PR. We can revisit the default name when we add support for parsing the type in the traditional order you wanted.

Reply:
Sounds good.